distributed/worker.py

```diff
@@ -1017,9 +1003,52 @@ async def start(self):
         enable_gc_diagnosis()
         thread_state.on_event_loop_thread = True
 
-        await self.listen(
-            self._start_address, **self.security.get_listen_args("worker")
+        ports = parse_ports(self._start_port)
+        for port in ports:
+            start_address = address_from_user_args(
+                host=self._start_host,
+                port=port,
+                interface=self._interface,
+                protocol=self._protocol,
+                security=self.security,
+            )
+            try:
+                await self.listen(
+                    start_address, **self.security.get_listen_args("worker")
+                )
+            except OSError as e:
+                if len(ports) > 1 and e.errno == errno.EADDRINUSE:
+                    continue
+                else:
+                    raise e
+            else:
+                self._start_address = start_address
+                break
+        else:
+            raise ValueError(
+                f"Could not start Worker on host {self._start_host}"
+                f"with port {self._start_port}"
+            )
```

Comment:
Does it make sense to try to move this logic into `Server.listen`?  It would mean changing the interface of that method to something like the following I think.

```python
async def listen(host, port, interface, protocol, security, ...)
```

Then maybe we can pull `address_from_user_args in, and apply this logic to the Scheduler and other future systems as well?

Reply:
(sorry to be late in reviewing by the way, if you want to just move forward and merge this that's fine with me too)

Reply:
For now I'd prefer to have this merged in as is, and then address the move to `Server.listen` in a follow-up PR

Reply:
+1 from me